### PR TITLE
Move CMake version check up

### PR DESCRIPTION
**Pull request by masterleinad**
Supersedes #4796 following https://github.com/kokkos/kokkos/pull/4789#discussion_r808610214. Also avoid redundantly setting `CMP0054` and `CMP0057` (except for KokkosConfig.cmake.in).

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,3 +1,6 @@
+IF(NOT KOKKOS_HAS_TRILINOS)
+  cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
+ENDIF()
```
Comment (dalg24) [inline CMakeLists.txt:3]: This won't work.  The variable `KOKKOS_HAS_TRILINOS` is not set at that point.
https://github.com/kokkos/kokkos/blob/2c2a8130d0ddabac4fe1649bc1a8c9a3364853b5/CMakeLists.txt#L34-L39

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,3 +1,4 @@
+cmake_minimum_required(VERSION 3.16 FATAL_ERROR)
```
Comment (dalg24) [inline CMakeLists.txt:1]: Plz confirm that this does not need to be guarded.
Looks good other than that.

**Comment by masterleinad** [comment]
`Cuda-9.2` failing with 
```
[  FAILED  ] default_exec.overlap_mdrange_policy
```
appears to be unrelated/spurious.

**Review by dalg24 — APPROVED** [review]

**Review by cz4rs — APPROVED** [review]

File: CMakeLists.txt (hunk 2 above)
Comment (henryiii) [reply CMakeLists.txt:1]: You can use `cmake_minimum_required(VERSION 3.16...3.22)` if you want to support a range - just make sure you test with 3.22. `FATAL_ERROR` is only needed if you want this to fail on CMake 2.6, which I'm pretty sure has completely vanished (and if someone is using 2.6, they have a lot of other problems!)

It is very much safe and recommended to place this at the top of files, see https://cliutils.gitlab.io/modern-cmake/chapters/basics.html

**Comment by PhilMiller** [comment]
Did this potentially also address #2503?

**Pull request merged**